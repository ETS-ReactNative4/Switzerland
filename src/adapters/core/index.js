export { default as useInterval } from './use-interval/index.js';
