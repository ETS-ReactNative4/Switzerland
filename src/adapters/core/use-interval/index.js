const nodes = new WeakMap();

/**
 * @function useInterval
 * ---
 * Re-renders the component every X milliseconds. Clears the interval on unmounting of the component
 * from the DOM.
 */
export default function useInterval({ node, lifecycle, render }) {
    return (milliseconds) => {
        switch (lifecycle) {
            case 'mount': {
                // Use the `setInterval` to re-render the component every X milliseconds.
                const interval = setInterval(render, milliseconds);
                nodes.set(node, interval);
                return interval;
            }

            case 'unmount':
                // Stop the interval when the node is unmounted from the DOM.
                clearInterval(nodes.get(node));
                break;
        }
    };
}
