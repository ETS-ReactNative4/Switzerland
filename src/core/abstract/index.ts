export abstract class Swiss {
    protected name: string;
    protected extend: null | string;

    constructor() {
        this.name = 'x-swiss';
        this.extend = null;
    }
}
