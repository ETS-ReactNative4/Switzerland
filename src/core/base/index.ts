import * as impl from '../impl';

export class Swiss implements impl.Swiss {
    async render(): Promise<HTMLElement> {
        return document.createElement('x-swiss');
    }
}
