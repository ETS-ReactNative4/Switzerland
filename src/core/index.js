import { getWindow, replaceTemplate } from '../utils.js';
import * as impl from './impl/index.js';
import * as utils from './utils.js';
import defaultController from './defaults/controller.js';
import defaultView from './defaults/view.js';

/**
 * @function create
 * ---
 * Takes the name of the web component and an array of functions that represent the middleware. Each
 * middleware item takes in the accumulated props, and yields props to pass to the next item in the list.
 */
export function create(name, { controller = defaultController, view = defaultView } = {}) {
    const [tag, constuctor, extend] = utils.parseName(name);

    try {
        window.customElements.define(tag, impl.client(constuctor, [controller, view]), extend && { extends: extend });
    } finally {
        return impl.server(extend ?? tag, [controller, view], extend ? tag : null);
    }
}

/**
 * @function render
 * ---
 * Takes the component tree and renders it to string for server-side rendering capabilities.
 */
export async function render(app, props = {}, options = { path: 'https://0.0.0.0/', root: '' }) {
    // Initialise the window on initial render so that it doesn't need to yield a promise every
    // single time it's invoked, instead the return value is memoized.
    await getWindow();

    // Render the app using the passed props.
    const node = await app.render(props, options);

    // JSDOM has an issue with appending children to the `template` node, so we merely find and replace
    // at this point to mimick the behaviour.
    return replaceTemplate(node.outerHTML);
}

/**
 * @function renderToStream
 * ---
 * Renders the component tree as usual but yields a readable Node stream that can be piped to the response.
 */
export async function renderToStream(app = {}, props, options = { path: 'https://0.0.0.0/', root: '' }) {
    const { Transform } = await import('stream');
    const stream = new Transform();

    // Push chunks of data into our stream.
    stream._transform = (chunk, _, done) => (this.push(chunk), done());

    async function run() {
        // Invoke the typical `render` function but with an attached stream, and end the stream once
        // the full component tree has been rendered.
        await render(app, props, { ...options, stream });
        stream.end();
    }

    return run(), stream;
}

/**
 * @function preload
 * ---
 * Collates the assets for server-side rendering so that the qualifying nodes can be placed in the head of the
 * document to preload which helps to prevent FOUC.
 */
export async function preload(...output) {
    const window = await getWindow();

    const links = output
        .flatMap((output) => output.match(/<link.+?>/gi))
        .map((link) => {
            if (!link) return;

            // Parse the link DOM string into a HTML node.
            const node = window.document.createElement('div');
            node.innerHTML = link;
            const child = node.firstChild;

            // Transform the link into a preload node for the head of the document.
            child.setAttribute('as', 'style');
            child.setAttribute('rel', 'preload');
            child.removeAttribute('key');
            child.removeAttribute('type');

            return child.outerHTML;
        });

    return links.join('\n');
}

/**
 * @function compose
 * ---
 * Allows the composition of controllers and views. In the case of controllers, the yielded objects are merged
 * together with duplicate props taking precendence in reverse order. In the case of views, the trees are taken
 * from left-to-right.
 */
export function compose(...fns) {
    return (props) => {
        const isViews = props.adapter == null;
        const newProps = fns.flatMap((fn) => fn(props));

        return isViews
            ? newProps
            : Object.values(newProps).reduce((accumProps, props) => ({ ...accumProps, ...props }), {});
    };
}
