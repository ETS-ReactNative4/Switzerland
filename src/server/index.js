import browserEnv from 'browser-env';
import { elements } from '../core/index.js';
import * as u from '../core/utils.js';
import { initialProps } from './utils.js';

browserEnv(['window', 'document', 'navigator']);

export async function render([name, middleware], attrs = {}, node = document.createElement(name)) {
    // Attach any custom attributes to the host element.
    Object.entries(attrs).forEach(([name, value]) => node.setAttribute(name, value));

    // Cycle through the middleware and append the "resolved" class name.
    const props = initialProps(node, middleware);
    await u.cycleMiddleware(node, props, middleware);
    node.classList.add('resolved');

    // Render recursively each custom child element.
    for (const name of elements.keys()) {
        const childNode = node.querySelector(name);
        const middleware = elements.get(name);
        childNode && (await render([name, middleware], {}, childNode));
    }

    return node.outerHTML;
}
