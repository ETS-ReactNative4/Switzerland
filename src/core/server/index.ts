import { DOMWindow } from 'jsdom';
import * as abstract from '../abstract';
import { getWindow } from '../../utils';
import type { Attributes } from './types';

export class Swiss extends abstract.Swiss {
    protected name: string;
    protected extend: null | string;

    constructor(name: string, extend?: string) {
        super();
        this.name = name;
        this.extend = extend ?? null;
    }

    async render(attributes: Attributes = {}): Promise<HTMLElement> {
        // Initialte the node and defined the `data-swiss` attribute which denotes a boundary.
        const window = (await getWindow()) as DOMWindow;
        const node = window.document.createElement(this.name);
        node.setAttribute('data-swiss', '');

        // Define the optional element to be extended from which will inherit its behaviours.
        if (this.extend != null) node.setAttribute('is', this.extend);

        // Define all of the attributes for the host node.
        for (const [key, value] of Object.entries(attributes)) node.setAttribute(key, value);

        return node;
    }
}

export function create(name: string, extend?: string): InstanceType<typeof Swiss> {
    return new Swiss(name, extend);
}
