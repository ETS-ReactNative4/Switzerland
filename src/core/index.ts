import * as server from './server';
import * as client from './client';
import { Attributes, Node, Children, View, Element } from '../types';
import { parseName, getRandomName } from './utils';

export function create(name: string, view: View = () => null): void | server.Swiss | CustomElementConstructor {
    const [tag, constructor, extend] = parseName(name);

    try {
        const component = client.create(constructor, view);

        window.customElements.define(
            window.customElements.get(tag) ? getRandomName() : tag,
            component,
            extend ? { extends: extend } : undefined
        );

        return component;
    } catch {
        return server.create(tag, extend, view);
    }
}

export async function render(component: server.Swiss): Promise<string> {
    const node = await component.render();

    return node.outerHTML;
}

export function h(element: Element, attributes: Attributes = {}, children: Children = []): Node {
    return { element, attributes, children };
}
