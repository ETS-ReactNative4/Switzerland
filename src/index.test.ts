import { create, render } from '.';
import * as server from './core/server';

describe.skip('create', () => {
    it('should be able to create server instances', () => {
        expect(create('x-imogen')).toBeInstanceOf(server.Swiss);
    });

    it('should be able to render server instances', async () => {
        const imogen = create('x-imogen', () => 'Hi Imogen!') as server.Swiss;
        expect(await render(imogen)).toMatchSnapshot();
    });
});
