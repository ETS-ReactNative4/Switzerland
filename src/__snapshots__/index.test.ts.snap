// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create should be able to render server instances 1`] = `"<x-imogen data-swiss=\\"\\">Hi Imogen!</x-imogen>"`;
